### ISSUE TITLE
Support for MSI Creator Z17 A12UGST

### ISSUE THREAD

**Issue by michaelbrylevskii** [BeardOverflow/msi-ec#180]
### Laptop model

MSI Creator Z17 A12UGST

### EC firmware version

17N1EMS1.109

### EC memory dump

```text
| _0 _1 _2 _3 _4 _5 _6 _7 _8 _9 _a _b _c _d _e _f
-----+------------------------------------------------
0x0_ | 00 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x1_ | 00 00 00 00 00 16 00 00 00 00 00 00 00 00 00 00
0x2_ | 00 00 00 00 00 00 00 00 0a 05 00 00 00 04 0b 0b
0x3_ | 03 01 00 0d 03 00 50 81 44 16 60 3b 3a 02 c0 00
0x4_ | 44 16 3d 00 1d 0f 00 00 12 09 ff 3b e0 0b f8 43
0x5_ | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x6_ | 00 00 00 00 00 00 00 00 25 00 37 40 49 4c 50 55
0x7_ | 64 28 28 2d 32 3c 46 4b 55 4b 08 03 03 03 02 03
0x8_ | 23 00 37 3d 43 48 4d 52 64 28 28 2d 32 3c 46 4b
0x9_ | 55 5f 0a 03 04 03 03 03 02 00 6e 02 00 66 24 00
0xa_ | 31 37 4e 31 45 4d 53 31 2e 31 30 39 30 36 30 36
0xb_ | 32 30 32 32 31 30 3a 35 30 3a 35 32 00 00 00 08
0xc_ | 00 00 07 25 00 00 00 00 00 b6 00 c9 00 b0 00 00
0xd_ | 00 00 c1 80 0d 00 05 bc 00 01 00 00 00 00 00 00
0xe_ | e2 00 00 1d 0f 01 00 40 01 00 00 00 00 b8 00 00
0xf_ | 00 00 70 00 24 3c 32 23 3c 32 00 00 00 00 00 00
```

### GPU

Nvidia

### Is your keyboard RGB?

Yes (multi color RGB)

**Comment by michaelbrylevskii** [comment]
There is the results of my EC research.

### Specs

**Laptop Model:** MSI Creator Z17 A12UGST
**EC:** 17N1EMS1

### Functions

**Microphone mute, with LED**
Address: `2C`
Values:
    - `00` - mic enabled (no mute, LED disabled)
    - `02` - mic disabled (mute, LED enabled)

**Audio mute, with LED**
Address: `2D`
Values:
    - `04` - sound enabled (no mute, LED disabled)
    - `06` - sound disabled (mute, LED enabled)

**Fn-lock (Fn+Esc), with LED**
Address: `D9`
Values:
    - `01` - custom actions mode (LED disabled)
    - `03` - function button mode (LED enabled)

**Camera toggle**
Address: `2E`
Values:
    - `08` - camera enabled
    - `09` - camera disabled

**Fn-Win Swap**
Address: `E8`
Values:
    - `01` - default (Win on left side)
    - `11` - swapped (Fn on left side)

**Battery thresholds**
Address: `D7`
Values:
    - `E4` - ? - AI Charger (I think values will change by software)
    - `E4` - Best for Mobility (100%)
    - `D0` - Balanced (70%-80%)
    - `BC` - Best for Battery (50%-60%)

**User scenario**
Address: `D2,D3,D4,EB` (I checked it a few times, and I don't understand their exact functions, `D2` look like performance mode, `D4` look like fan mode)
Values:
    - `C1,83,0D,00`- ? - AI Smart Auto (I think values will change by software)
    - `C0,83,0D,00` - Hight Performance, fan auto
    - `C1,83 or 80,0D,00` - Balanced
    - `C1,83,1D,00` - Silent
    - `C2,80,0D,0F` - Battery Safe

**Fan mode (allowed only in Hight Performance mode)**
Address: `D4,98`
Values:
    - `0D,02` - Auto
    - `0D,82` - Cooler Boost
    - `4D,02` - Advanced
    - - Addresses of CPU fan speeds for thresholds: `72,73,74,75,76,77`
    - - Address of current target CPU fan speed: `71` (value is one from `72,73,74,75,76,77`)
    - - Addresses of GPU fan speeds for thresholds: `8A,8B,8C,8D,8E,8F`
    - - Address of current target GPU fan speed: `89` (value is one from `8A,8B,8C,8D,8E,8F`)
    - - Values:
    - - - `96` - !50%
    - - - `64` - !00%
    - - - `32` - 50%
    - - - `19` - 25%
    - - - `14` - 20%
    - - - `0A` - 10%

**Related with charging cable**
Address: `30,31`
Values:
    - `03,09` - Plugged
    - `02,0D` - Unplugged

**Keyboard backlight brightness**
No reaction (or maybe I didn't notice)

**Touchpad toggle**
No reaction (or maybe I didn't notice)

**Gaugers**
    - `68` - Locks like CPU temp
    - `80` - Locks like GPU temp (value `00` if GPU inactive)

**Comment by mutchiko** [comment]
Question: did you follow the [support guide](https://github.com/BeardOverflow/msi-ec/blob/main/docs/device_support_guide.md) when discovering values and addresses?

**Comment by glpnk** [comment]
Thanks for your research! Will you commit your findings to the project?

**Comment by michaelbrylevskii** [comment]
> Question: did you follow the [support guide](https://github.com/BeardOverflow/msi-ec/blob/main/docs/device_support_guide.md) when discovering values and addresses?

Yes, I did it on windows. I can attach a few different dumps (RWEverething) If it will useful.

**Comment by glpnk** [comment]
I think you posted enough data

Does keyboard switch mode/brightness on Linux, when keyboard mode/brightness button is pushed (if you have one)

**Comment by michaelbrylevskii** [comment]
> Thanks for your research! Will you PR your findings to the project?

I think I can, but it will takes a while, I'm new to Linux :)

**Comment by glpnk** [comment]
> I'm new to Linux :)

Nice, I hope you don't have weird problems with hardware, like broken GPU or slow charger detection

**Comment by michaelbrylevskii** [comment]
> Does keyboard switch mode/brightness on Linux, when keyboard mode/brightness button is pushed (if you have one)

Yes, I have keyboard brightness button, it works even on Linux, but without changes in EC memory. Maybe it works without EC functions.

**Comment by michaelbrylevskii** [comment]
> Can you share output of `lsusb` command?

```
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 003 Device 002: ID 5986:213a Bison Electronics Inc. FHD Camera
Bus 003 Device 003: ID 1038:113a SteelSeries ApS SteelSeries KLC
Bus 003 Device 004: ID 27c6:6094 Shenzhen Goodix Technology Co.,Ltd. Goodix USB2.0 MISC
Bus 003 Device 005: ID 09da:3711 A4Tech Co., Ltd. USB Device
Bus 003 Device 006: ID 0eef:c003 D-WAV Scientific Co., Ltd eGalaxTouch P86H82 2666 v03_03 k4.16.190
Bus 003 Device 007: ID 8087:0033 Intel Corp. AX211 Bluetooth
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 004 Device 002: ID 090c:1000 Silicon Motion, Inc. - Taiwan (formerly Feiya Technology Corp.) Flash Drive
```

Two of them are externally connected devices:
```
Bus 003 Device 005: ID 09da:3711 A4Tech Co., Ltd. USB Device
Bus 004 Device 002: ID 090c:1000 Silicon Motion, Inc. - Taiwan (formerly Feiya Technology Corp.) Flash Drive
```

**Comment by glpnk** [comment]
Your keyboard is USB, so you can try to use [OpenRGB](https://github.com/CalcProgrammer1/OpenRGB) to control it, if it is supported.

**Comment by michaelbrylevskii** [comment]
@glpnk, I've done it, it works good, I've tested all existed features. 
PR: https://github.com/BeardOverflow/msi-ec/pull/181

If I understand correctly, now I can't read/write fan speed values for advanced fan mode (only viw debug), I'm right?

**Comment by glpnk** [comment]
@michaelbrylevskii Hi, custom fan curve is not supported yet
